File: merge_sort.js
/*
 * Javascript implementation for merge sort, complexity O(NlogN)
 */

var merge = function(L, R) {
    var A = L.concat(R);
    L.push(Infinity);
    R.push(Infinity);

    var i = j = 0;
    for (var k = 0; k < A.length; k++) {
        if (L[i] <= R[j]) {
            A[k] = L[i];
            i++;
        } else {
            A[k] = R[j];
            j++;
        }
    }

    return A;
};

var merge_sort = function(A) {
    if (A.length > 1) {
        var q = parseInt(A.length/2);
        var L = A.slice(0, q);
        var R = A.slice(q);
        L = merge_sort(L);
        R = merge_sort(R);
        return merge(L, R);
    } else {
        return A;
    }
};

var A = [3, 41, 52, 26, 38, 57, 9, 49];
console.log("Before sort: " + A.toString())
var B = merge_sort(A)
console.log("After sort: " + B.toString())

File: merge_sort.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Python implementation for merge sort, complexity O(NlogN)

def merge(A, p, q, r):
    L = A[p: q+1]
    R = A[q+1: r+1]
    L.append(float('inf'))
    R.append(float('inf'))

    i, j = 0, 0
    for k in range(p, r+1):
        if L[i] <= R[j]:
            A[k] = L[i]
            i += 1
        else:
            A[k] = R[j]
            j += 1

def merge_sort(A, p, r):
    if p < r:
        q = (p+r)//2
        merge_sort(A, p, q)
        merge_sort(A, q+1, r)
        merge(A, p, q, r)

if __name__ == '__main__':
    A = [3, 41, 52, 26, 38, 57, 9, 49]
    print("Before sort:{}".format(A))
    merge_sort(A, 0, 7)
    print("After sort:{}".format(A))
    
